src/internal.rs

```diff
@@ -928,89 +931,158 @@ impl BitDepthDependentType for AlPal {
+#[repr(C, align(64))]
+pub struct ScratchLapInter([u8; SCRATCH_COMPINTER_SIZE]);
 
-#[derive(Clone, Copy)]
-#[repr(C)]
-pub union Rav1dTaskContext_scratch_lap {
-    pub lap: BitDepthUnion<Lap>,
-    pub c2rust_unnamed: Rav1dTaskContext_scratch_compinter_seg_mask,
+impl ScratchLapInter {
+    pub fn lap_mut<BD: BitDepth>(&mut self) -> &mut [BD::Pixel; 128 * 32] {
+        FromBytes::mut_from_prefix(&mut self.0).unwrap()
+    }
+
+    pub fn inter_mut(&mut self) -> &mut ScratchCompinter {
+        FromBytes::mut_from_prefix(&mut self.0).unwrap()
+    }
 }
 
+const EMU_EDGE_LEN: usize = 320 * (256 + 7);
 // stride=192 for non-SVC, or 320 for SVC
-pub struct EmuEdge;
+#[derive(Clone, Copy, FromZeroes, FromBytes, AsBytes)]
+#[repr(C, align(32))]
+pub struct ScratchEmuEdge([u8; EMU_EDGE_LEN * 2]);
 
-impl BitDepthDependentType for EmuEdge {
-    type T<BD: BitDepth> = [BD::Pixel; 320 * (256 + 7)];
+impl ScratchEmuEdge {
+    pub fn buf_mut<BD: BitDepth>(&mut self) -> &mut [BD::Pixel; EMU_EDGE_LEN] {
+        FromBytes::mut_from_prefix(&mut self.0).unwrap()
+    }
 }
 
-#[derive(Clone, Copy)]
+#[derive(Clone, Copy, FromZeroes, FromBytes, AsBytes)]
 #[repr(C)]
-pub struct Rav1dTaskContext_scratch_lap_emu_edge {
-    pub c2rust_unnamed: Rav1dTaskContext_scratch_lap,
-    pub emu_edge: BitDepthUnion<EmuEdge>,
+pub struct ScratchInter {
+    pub lap_inter: ScratchLapInter,
+    pub emu_edge: ScratchEmuEdge,
 }
 
-#[derive(Clone, Copy)]
-#[repr(C)]
-pub struct Rav1dTaskContext_scratch_pal {
+#[derive(Clone, Copy, FromZeroes, FromBytes, AsBytes)]
+#[repr(C, align(64))]
+pub struct ScratchPal {
     pub pal_order: [[u8; 8]; 64],
     pub pal_ctx: [u8; 64],
 }
 
-#[derive(Clone, Copy)]
-#[repr(C)]
-pub union Rav1dTaskContext_scratch_levels_pal {
-    pub levels: [u8; 1088],
-    pub c2rust_unnamed: Rav1dTaskContext_scratch_pal,
+#[derive(Clone, Copy, FromZeroes, FromBytes, AsBytes)]
+#[repr(C, align(64))]
+pub struct ScratchLevelsPal([u8; 1088]);
+
+impl ScratchLevelsPal {
+    pub fn levels_mut(&mut self) -> &mut [u8; 1088] {
+        &mut self.0
+    }
+
+    pub fn pal_mut(&mut self) -> &mut ScratchPal {
+        ScratchPal::mut_from_prefix(&mut self.0).unwrap()
+    }
 }
 
-#[derive(Clone, Copy)]
-#[repr(C)]
-pub struct InterIntraEdgePalBD<BD: BitDepth> {
-    pub interintra: [BD::Pixel; 64 * 64],
-    pub edge: [BD::Pixel; 257],
-    _align: [BD::AlignPixelX8; 0],
-    pub pal: [[BD::Pixel; 8]; 3], /* [3 plane][8 palette_idx] */
+#[derive(Clone, Copy, FromZeroes, FromBytes, AsBytes)]
+#[repr(C, align(64))]
+pub struct ScratchInterintraBuf([u8; 64 * 64 * 2]);
+
+impl ScratchInterintraBuf {
+    pub fn buf_mut<BD: BitDepth>(&mut self) -> &mut [BD::Pixel; 64 * 64] {
+        FromBytes::mut_from_prefix(&mut self.0).unwrap()
+    }
 }
 
-pub struct InterIntraEdgePal;
+#[derive(Clone, Copy, FromZeroes, FromBytes, AsBytes)]
+#[repr(C, align(32))]
+pub struct ScratchEdgeBuf([u8; 257 * 2 + 30]); // 257 Pixel elements + 30 padding bytes
```

Comment:
What do you mean by "return the 30 padding bytes"? Return the full size of the buffer, including padding, from `buf_mut`? I don't think doing that would be a good idea, the buffer had a specific length for a reason and we shouldn't be changing the size of it. There's also code using this data that assumes a length of 257.

Reply:
So we start with a `[u8; 257 * 2 + 30]`.  Say we then return a `&[u16; 257]` for bpc16, and then asm accesses 30 bytes after that.  That violates stacked borrows since it expands provenance.  So to do things correctly with no UB, we need to keep that 30 bytes there the whole time.  Is it worth it?  I'm not sure, but it technically is UB.

Reply:
> Given that the original code was written to be a buffer of 257 elements without any padding at the end, I think it's reasonably safe to assume that the asm won't read beyond that

I'm pretty sure the padding is there since asm is using SIMD instructions that over-read.

> If it is, the behavior isn't going to be any different than it was in the original C.

C has different UB rules than Rust, and doing it in Rust with all raw pointers is also different than creating that intermediate slice.